codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -579,102 +318,85 @@ public boolean isEmpty() {
+        }
         return this;
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public Headers<T> set(T name, T value) {
-        name = convertName(name);
+        validateName(name);
         checkNotNull(value, "value");
-        int h = hashCodeGenerator.generateHashCode(name);
-        int i = index(h);
-        remove0(h, i, name);
-        add0(h, i, name, value);
+        Object oldValue = map.put(name, value);
+        updateSizeAfterSet(oldValue, 1);
         return this;
     }
 
     @Override
     public Headers<T> set(T name, Iterable<? extends T> values) {
-        name = convertName(name);
+        validateName(name);
         checkNotNull(values, "values");
-
-        int h = hashCodeGenerator.generateHashCode(name);
-        int i = index(h);
-        remove0(h, i, name);
-        for (T v : values) {
-            if (v == null) {
-                break;
-            }
-            add0(h, i, name, v);
+        List<T> list = newList();
+        for (T value : values) {
+            list.add(value);
         }
-
+        Object oldValue = map.put(name, list);
+        updateSizeAfterSet(oldValue, list.size());
         return this;
     }
 
     @Override
     public Headers<T> set(T name, T... values) {
-        name = convertName(name);
+        validateName(name);
         checkNotNull(values, "values");
-
-        int h = hashCodeGenerator.generateHashCode(name);
-        int i = index(h);
-        remove0(h, i, name);
-        for (T v : values) {
-            if (v == null) {
-                break;
-            }
-            add0(h, i, name, v);
+        List<T> list = newList(values.length);
+        for (int i = 0; i < values.length; i++) {
+            list.add(values[i]);
         }
-
+        Object oldValue = map.put(name, list);
+        updateSizeAfterSet(oldValue, values.length);
         return this;
     }
 
     @Override
     public Headers<T> setObject(T name, Object value) {
-        return set(name, valueConverter.convertObject(checkNotNull(value, "value")));
+        checkNotNull(value, "value");
+        T convertedValue = checkNotNull(valueConverter.convertObject(value), "convertedValue");
+        return set(name, convertedValue);
     }
 
     @Override
     public Headers<T> setObject(T name, Iterable<?> values) {
-        name = convertName(name);
+        validateName(name);
         checkNotNull(values, "values");
-
-        int h = hashCodeGenerator.generateHashCode(name);
-        int i = index(h);
-        remove0(h, i, name);
-        for (Object o : values) {
-            if (o == null) {
-                break;
-            }
-            T converted = valueConverter.convertObject(o);
-            checkNotNull(converted, "converted");
-            add0(h, i, name, converted);
+        List<T> list = newList();
+        for (Object value : values) {
+            T convertedValue = checkNotNull(valueConverter.convertObject(value), "convertedValue");
```

Comment:
Should we `checkNotNull(value, "")` before giving it to `convertObject`?  I think we are doing this elsewhere.  Perhaps we should document that `null` will not be passed in to this method?